Add fhir resource get view

mkangia opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Adds the view to GET a resource in FHIR format.
1. The URL is expected to be `.../fhir/R4/Patient/{case_id}`
2. Authentication and authorization is still not configured so putting this available only for superusers for now so that this can be QA'ed. QA would happen on staging but should be safe as well if this gets deployed. It's behind a domain-based feature flag and then only available to superusers

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`FHIR_INTEGRATION`

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This adds a new URL but has authentication added. The access is only for superusers + for domains with feature flag enabled, so does not put risk of data getting accessed by unwanted users. Additionally also needs Data dictionary set up, else it just returns a 400, which too is behind an additional feature flag (`DATA_DICTIONARY`)

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

kaapstorm committed:
FHIRRepeater, form, views

kaapstorm committed:
test_repeater_helpers

kaapstorm committed:
Add jsonschema to requirements

kaapstorm committed:
Validate resources

kaapstorm committed:
DRY get_case_trigger_infos()

kaapstorm committed:
deepmerge: Skip case if it has no data

kaapstorm committed:
Register patients

kaapstorm committed:
End-to-end test to register patient

kaapstorm committed:
Supercase/subcase test

kaapstorm commented on corehq/motech/fhir/urls.py:
```diff
@@ -0,0 +1,7 @@
+from corehq.motech.fhir.views import get_view
+
+urlpatterns = [
+    url(r'^fhir/R4/(?P<resource_type>\w+)/(?P<resource_id>\w+)$', get_view, name="fhir_get_view"),
```
You'll want to grab the FHIR version as well. You could hard-code it while we only support one version, but it would be better to get it from the URL. You can validate against `fhir.const.FHIR_VERSIONS`.

mkangia replied:
Done

kaapstorm commented on corehq/motech/fhir/views.py:
```diff
@@ -0,0 +1,29 @@
+    except CaseNotFound:
+        return JsonResponse(status=400, data={'message': f"Could not find resource with ID {resource_id}"})
+
+    if case.type not in (FHIRResourceType.objects.filter(domain=domain, name=resource_type).
+                         values_list('case_type__name', flat=True)):
```
```suggestion
    fhir_version = [v[0] for v in FHIR_VERSIONS if v[1] == version_name][0]
    if not FHIRResourceType.objects.filter(
        domain=domain,
        name=resource_type,
        fhir_version=fhir_version,
        case_type__name=case.type,
    ).exists():
```

mkangia replied:
could not commit directly from here since the view arguments would need to be changed as well, so made a commit.
Thanks for reminding of "exists"

kaapstorm commented on corehq/motech/fhir/views.py:
```diff
@@ -0,0 +1,29 @@
+    if case.type not in (FHIRResourceType.objects.filter(domain=domain, name=resource_type).
+                         values_list('case_type__name', flat=True)):
+        return JsonResponse(status=400, data={'message': "Invalid Resource Type"})
+    response = {
+        'resourceType': resource_type,
+        'id': resource_id
+    }
+    response.update(build_fhir_resource(case))
```
This should end up as
```suggestion
    response = get_fhir_resource_for_case(case, fhir_version)
```

mkangia replied:
are those added by `build_fhir_resource` now?

kaapstorm replied:
Added in [this commit](https://github.com/dimagi/commcare-hq/pull/29261/commits/b0ee0d495f3b1687f64b96343427963e5c6f8bf4#diff-433e7032abf02a0c78231568078bff8181998607bf011edb22ea4b1d3018041fR199-R200), in PR #29261

mkangia replied:
Cool, that PR is still in draft state, and this one is in QA. So going to keep this as is for now and make a note to update this later.

kaapstorm reviewed (commented):
This is looking great.

kaapstorm committed:
Serializers: Support splitting strings into lists

kaapstorm committed:
Build out `build_fhir_resource()`

kaapstorm committed:
Build out repeater_helpers

mkangia commented:
thanks for the first pass @kaapstorm , addressed 2/3 points and replied on 1.

millerdev committed:
Remove ModelActionAudit

This has not been used for years. The main setting, AUDIT_MODEL_SAVE, was never set properly (the paths were invalid; see https://github.com/dimagi/commcare-hq/commit/2af2ade5ecee4305127284cd219f67cbb5e458a5 ), so it had no effect. User model auditing was enabled by default from https://github.com/dimagi/commcare-hq/commit/fcc9bfb122448f62f329f295e40b3817db2a3e74 (2011) until https://github.com/dimagi/commcare-hq/commit/ffa647228f89aae8e90101b030190d528eadc420 (2014). Data on production corresponds with those findings.

czue committed:
add fixture data if it exists

czue committed:
make code more resilient to bad data

czue committed:
include fixtures in application api

AmitPhulera committed:
Merge pull request #29263 from dimagi/ap/fix-filters-in-case-list-report

Fix filters in Case List Report

stephherbers committed:
added output-file optional parameter to run all command and creates new file

stephherbers committed:
adds options output-file parameter to updates cases for the rest of the scripts and adds a logging function to open the file and write input

stephherbers committed:
adds logging capability to individual scripts

kaapstorm committed:
Delete in chunks, every day

orangejenny committed:
Removed soft assert

kaapstorm committed:
Clean up tests

orangejenny committed:
Updated enterprise permissions page to disallow a mirror whose source is itself

gherceg committed:
adress feedback part 2

consolidate tests, better documentation

gherceg committed:
thanks stickler

orangejenny committed:
Merge pull request #27598 from dimagi/jls/migrateCommtrackConfig-2

Couch to SQL migration: CommtrackConfig (2 of 3)

orangejenny committed:
Merge pull request #29271 from dimagi/jls/self-referential-enterprise-permissions

Updated enterprise permissions page to disallow a mirror whose source…

gherceg committed:
pollish export settings UI

use new widget for aligned checkbox, add labels

stephherbers committed:
Update staging.yaml

kaapstorm committed:
Tests use SQL backend

biyeun committed:
Merge pull request #29267 from dimagi/bmb/SAAS-11902

change_plan does not support start_date

gherceg committed:
bad typo

orangejenny committed:
Merge pull request #29270 from dimagi/jls/remote-soft-assert

Removed soft assert

gherceg committed:
Merge pull request #29273 from dimagi/gh/improve-odata-text

Pollish Export Settings UI

kaapstorm committed:
Move test utils to module with related test utils

mjriley committed:
Merge pull request #29266 from dimagi/mjr/xss_misc

kaapstorm committed:
Merge pull request #29250 from dimagi/nh/fhir/rep-2

FHIRRepeater part 2

mjriley committed:
Merge pull request #29245 from dimagi/mjr/xss_locations

mjriley committed:
Merge pull request #29265 from dimagi/mjr/xss_settings

removed mark_safe usages in settings

biyeun committed:
add a method to IdentityProvider to return a list of active domains/projects associated with that Identity Provider

biyeun committed:
add a method to IdentityProvider to return the identity provider for a given username (if it exists)

biyeun committed:
remove unused import

biyeun committed:
add a method to sso test generator to quickly create an enterprise plan for tests

biyeun committed:
add utility to check if a request is using sso

biyeun committed:
add a utility to check if a request is blocked by unmet sso requirements

biyeun committed:
add tests for request helper utilities

Commit:
'updating files: scripts/staging.yaml'

'[ci skip]'

biyeun committed:
Add functionality to login_and_domain decorator that blocks the request if the user is logged in with SSO and does not meet the SSO requirements for viewing that domain
NOTE: This is carefully feature-flagged for ENTERPRISE_SSO users right now.

orangejenny committed:
Revert "Web Apps: Use SelectWoo, a more accessible fork of Select2, for comboboxes"

gherceg committed:
Merge pull request #29244 from dimagi/gh/last-path-component

Add additional find_question_id check for last path component

biyeun committed:
add domain decorators.py as high risk

biyeun committed:
Merge pull request #29278 from dimagi/bmb/update-high-risk-flag

mark domain decorators.py as high risk

orangejenny committed:
Merge pull request #29277 from dimagi/revert-29022-cdg/accessible_combobox2

Revert "Web Apps: Use SelectWoo, a more accessible fork of Select2, for comboboxes"

millerdev committed:
Merge pull request #29274 from dimagi/dm/rm-modelactionaudit

Remove ModelActionAudit and associated features

kaapstorm committed:
Fix teardown

snopoke committed:
update docker-compose files for formplayer

snopoke committed:
increase logging in docker container

snopoke committed:
create kafka topics in bootstrap

snopoke committed:
upgrade kafka

snopoke committed:
separate formplayer URL for webapps

Webapps is accessing formplayer from the browser which is outside the docker
container and can't access `formplayer:8080`.

snopoke committed:
rename setting

kaapstorm committed:
Tests

kaapstorm committed:
Merge pull request #29258 from dimagi/nh/motech/relations

ValueSource supports case relationships

czue committed:
update test with desired behavior

czue committed:
add get_external_instances and test

snopoke committed:
move env vars to base compose file

snopoke committed:
move KAFKA_ADVERTISED_HOST_NAME to root compose file

czue committed:
add extract_instance_id_from_nodeset and test

czue committed:
update test api and make it work

czue committed:
make string raw

millerdev committed:
corehq/{ex-submodules -> apps}/auditcare

stephherbers committed:
Merge pull request #29240 from dimagi/smh/fix-blank-inactive-run-all-script

Added condition if a domain does not have a traveler active location

kaapstorm committed:
Support "coordinate" and "geometry" in enrollments

orangejenny committed:
Merge pull request #29256 from dimagi/jls/dd-upload-error-handling-improvements

Data dictionary upload error handling improvements

biyeun committed:
don't use the word "test" in a test utility

stephherbers committed:
Merge branch 'master' into smh/run-all-command-logging

gherceg committed:
Merge pull request #29182 from dimagi/dependabot/pip/requirements/python-magic-0.4.22

Bump python-magic from 0.4.18 to 0.4.22 in /requirements

gherceg committed:
Merge pull request #29213 from dimagi/dependabot/pip/requirements/django-extensions-3.1.1

Bump django-extensions from 3.1.0 to 3.1.1 in /requirements

gherceg committed:
Update staging.yaml

stephherbers committed:
added output_file as parameter for run_command

stephherbers committed:
added so case error aslo prints which domain it's associated with

stephherbers committed:
cleaned up commands but condensing some using kwargs

millerdev committed:
Adjust imports, etc. for moved auditcare

orangejenny committed:
Updated USH script to fix all "'patient'" cases

orangejenny committed:
Moved case property template to separate HTML file

biyeun committed:
add utility for getting an email domain from a username

orangejenny committed:
Merge pull request #29283 from dimagi/orangejenny-patch-6

Updated USH script to fix all "'patient'" cases

orangejenny committed:
Added sortability

biyeun committed:
refactor how domain trusts the identity provider

biyeun committed:
quickcache these methods as they will be called on each request.

biyeun committed:
note to self so i don't forget

kaapstorm committed:
Merge pull request #29281 from dimagi/nh/dhis2/schema

DHIS2: Support "coordinate" and "geometry" in enrollments

biyeun committed:
pr feedback

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

biyeun committed:
update docstring

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

biyeun committed:
renders a more user-friendly view for users viewing a domain's page if the domain has not yet trusted the identity provider.

orangejenny committed:
Updated sticky search to work with daterange inputs

orangejenny committed:
Added translation

orangejenny committed:
Consolidated type-specific logic

orangejenny committed:
Fixed whitespace

orangejenny committed:
Update staging.yaml

orangejenny committed:
Fixed white space some more

orangejenny committed:
Update staging.yaml

biyeun committed:
make sure we clear the quickcache at the tear down of each test

orangejenny committed:
Update staging.yaml

orangejenny committed:
Update staging.yaml

calellowitz committed:
add wiki link

calellowitz committed:
add missing comma

orangejenny committed:
Update staging.yaml

calellowitz committed:
Merge pull request #29289 from dimagi/ce/phone-doc-link

add wiki link

orangejenny committed:
Update staging.yaml

orangejenny committed:
Updated linked domains to fail when downstream app reference is ambiguous

czue committed:
update API key to "data_source"

kaapstorm committed:
Merge branch 'master' into nh/motech/ghost

kaapstorm committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

kaapstorm committed:
Merge pull request #29262 from dimagi/nh/motech/ghost

Raw-delete RequestLog

kaapstorm committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

mkangia committed:
add max elgible counts in a month

mkangia committed:
add female death type to DS

mkangia committed:
fix name for template

mkangia committed:
add new report

millerdev committed:
Merge pull request #29282 from dimagi/dm/mv-auditcare

Move auditcare from ex-submodules to corehq.apps

mkangia committed:
isort

biyeun committed:
Merge pull request #29276 from dimagi/bmb/sso-11780

[SSO] Block users from viewing Domain if logged in with SSO and Domain does not Trust IdP

orangejenny committed:
Merge pull request #29290 from dimagi/jls/linked-domain-fail-multiple-downstream

Updated linked domains to fail when downstream app reference is ambig…

orangejenny committed:
Merge pull request #29285 from dimagi/jls/sortable-case-claim-properties

Make case claim properties sortable

orangejenny committed:
Removed old band-aid from build_requirejs

kaapstorm committed:
Format JSON request body in a readable way

snopoke committed:
Merge pull request #29280 from dimagi/sk/formpalyer-docker

formpalyer docker revamp

kaapstorm committed:
Update tests

czue committed:
Merge pull request #29269 from dimagi/cz/app-fixtures-api

add fixture information to questions in Application API

kaapstorm committed:
Don't render empty request body as "None"

stephherbers committed:
output_file not added to extra_options

stephherbers committed:
opens file with 'a' param

mkangia committed:
Merge pull request #29294 from dimagi/mk/india-nutrition-reports/hcm-thr-fix

Add max elgible counts in a month

stephherbers committed:
updated name of failed_cases to errors and added more details to the error messages

kaapstorm committed:
Merge pull request #29298 from dimagi/nh/dhis2/request_body

Format request body

mkangia committed:
simply return 404

also ensures that the case belongs to the domain

mkangia committed:
Merge pull request #29295 from dimagi/mk/india-nutrition-reports/delivery-deaths

Add report for delivery deaths

biyeun committed:
temporarily remove datetimepicker from requirejs files while we figure out weird rjs bug

proteusvacuum committed:
raise CaseNotFound if domain is mismatched

biyeun committed:
Merge pull request #29301 from dimagi/bmb/temp-remove-datetimepicker

temporarily remove datetimepicker from requirejs files while we figure out weird rjs bug

orangejenny committed:
Update staging.yaml

mkangia committed:
add missed column in data source

gherceg committed:
remove unnecessary block

mkangia committed:
Merge pull request #29303 from dimagi/mk/india-nutrition-reports/delivery-deaths-fix

add missed column in data source

proteusvacuum committed:
Merge pull request #29302 from dimagi/fr/case-domain-mismatch

raise CaseNotFound if domain is mismatched

orangejenny committed:
Update staging.yaml

orangejenny committed:
Update staging.yaml

AmitPhulera committed:
add check to include locaions while creating case list es query

AmitPhulera committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

AmitPhulera committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

AmitPhulera committed:
add memoized back

czue committed:
move test sms generator to standalone module

czue committed:
add management command to generate fake sms data

czue committed:
add randomize flag and use recent dates if randomize enabled

czue committed:
randomize enums

czue committed:
the stickler never sleeps

gherceg committed:
Merge pull request #29305 from dimagi/gh/remove-duplicate-query-string

Remove unnecessary block that adds filters and date_range again

czue committed:
add messaging event resource

biyeun committed:
Merge pull request #29297 from dimagi/bmb/sso-11780-ui

[SSO] Add user-friendly view when a user tries to access domain that doesn't trust IdP

czue committed:
Merge pull request #29307 from dimagi/cz/sms-data-generator

Add basic sms data generator

AmitPhulera committed:
Merge pull request #29306 from dimagi/ap/fix-location-filtering-in-case-list-report

Check for locations also while creating case list es query

biyeun committed:
cleanup remnants of the old failed datetimepicker

biyeun committed:
use new datetimepicker widget for bootstrap 3 (note the old one is still used in webapps and i'm not touching it yet)

biyeun committed:
use new datetimepicker widget that is hopefully requirejs friendly

biyeun committed:
add custom styling for datetimepicker widget because the default styling really sux

biyeun committed:
lint

stephherbers committed:
added start and end time of script run to the end of the output file

orangejenny committed:
Added handling for free text dates

orangejenny committed:
Update staging.yaml

orangejenny committed:
Fixed lint errors

orangejenny committed:
Fixed whitespace

biyeun committed:
Merge pull request #29310 from dimagi/bmb/new-datetime-picker

new datetimepicker (cleanup old datetimepicker EXCEPT in webapps)

orangejenny committed:
Update staging.yaml

orangejenny committed:
Update staging.yaml

orangejenny committed:
Merge pull request #29264 from dimagi/jls/rename-master-id

Renamed Application.master_id => origin_id

kaapstorm committed:
Refactor tests for Couch, add for SQL

kaapstorm committed:
Merge pull request #29249 from dimagi/nh/fhir/rep-1

FHIRRepeater part 1

kaapstorm committed:
Merge pull request #29260 from dimagi/nh/fhir/model_changes

Alter FHIRResourceType

kaapstorm committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

kaapstorm committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

millerdev committed:
Remove ineffective admin site login/logout patches

Ineffective because `admin.site.urls`, which references `admin.site.login`, is calculated and installed in the base urls.py before these patches were applied. Tested locally and verified that these were not called. One reason it is safe to remove this is because django-two-factor-auth overrides the admin login pages with a redirect to the main login page.

millerdev committed:
Remove unused auth_login/auth_logout views

millerdev committed:
Add tests and fix bug in AuditMiddleware

view_func.func_name was removed in Python 3

millerdev committed:
Refactor AuditMiddleware

millerdev committed:
Refactor access audit methods

stephherbers committed:
added newline character for better readability

mkangia committed:
allow access via api key as well

mkangia committed:
add new templates

czue committed:
allow ordering by date (necessary for DET pagination)

gherceg committed:
Merge pull request #29152 from dimagi/dependabot/pip/requirements/django-compressor-2.4

Bump django-compressor from 2.1 to 2.4 in /requirements

gherceg committed:
Update staging.yaml

kaapstorm committed:
Case type instances are unique but not their names

mkangia committed:
Merge pull request #29314 from dimagi/mk/india-nutrition-reports/v2_templates_1

Add new sum when templates

millerdev committed:
Add auditcare tests for login/logout

Replaces the tests deleted in earlier commit "Remove unused auth_login/auth_logout views"

orangejenny committed:
Merge pull request #29287 from dimagi/jls/web-apps-sticky-date-range

Updated sticky search to work with daterange inputs

orangejenny committed:
Update staging.yaml

orangejenny committed:
Merge pull request #29296 from dimagi/jls/requirejs-remove-bandaid

Removed old band-aid from build_requirejs

mjriley committed:
Removed mark_safe usages in translations

orangejenny committed:
Merge pull request #29116 from dimagi/jls/remove-autolaunch

Removed autolaunch flag from web apps

calellowitz committed:
clear page on query

calellowitz committed:
also clear search bar

millerdev committed:
Disable broken audit_login_failed signal handler

Failed login logging is already being done elsewhere, so probably makes sense wait to fix this until we decide if that will be merged into auditcare.

calellowitz committed:
Merge pull request #29316 from dimagi/ce/search-page

clear page on query

biyeun committed:
add method for clearing the domain cache

biyeun committed:
add method for establishing trust between the current identity provider and a domain

biyeun committed:
ensure that a trust is created when a user visits their own domain for the first time using SSO and display a message to the user

biyeun committed:
add utility for properly populating the request with a session

biyeun committed:
modify test for ensuring that a TrustedIdentityProvider relationship is created for non-enterprise domains owned by an sso-logged in user

shubham1g5 committed:
Shows notification for EOF navigation

czue committed:
Merge branch 'master' into cz/message-event-api

czue committed:
add some tests for messaging event API

czue committed:
lint

mkangia committed:
add branch for staging

millerdev committed:
Improve test and assert function names

millerdev committed:
assertEquals is deprecated

stephherbers committed:
removed unnecessary file close calls

millerdev committed:
fqview -> view_name

millerdev committed:
Refactor AuditMiddleware to new-style middleware

millerdev committed:
Improve defaults for make_view() test helper

millerdev committed:
Improve audit assertion failure output

czue committed:
Merge pull request #29308 from dimagi/cz/message-event-api

add basic messaging event API

millerdev committed:
Improve default view parameters in tests

millerdev committed:
make_middleware -> configured_middleware

millerdev committed:
Move request into setUp

shubham1g5 committed:
Merge pull request #29319 from dimagi/eofNotificationFix

Shows notification message for EOF navigation

millerdev committed:
Add tests for AuditMiddleware response processing

kaapstorm committed:
Fix repeater view URLs

orangejenny committed:
Update staging.yaml

orangejenny committed:
Update staging.yaml

orangejenny committed:
Update staging.yaml

stephherbers committed:
Merge pull request #29284 from dimagi/smh/run-all-command-logging

Add option for output file for logging for the casedb scripts

millerdev committed:
Use explicit class-based view name

millerdev committed:
Merge pull request #29315 from dimagi/dm/auditcare-cleanup

auditcare cleanup

biyeun committed:
Merge pull request #29318 from dimagi/bmb/sso-auto-trust-self-domain

[SSO] Alert created_by user on first time visiting a Domain that recently auto-trusted their new IdP

orangejenny committed:
Fixed queryData not getting cleared

orangejenny committed:
Update staging.yaml

kaapstorm committed:
Raise a better error

kaapstorm committed:
Don't cast None as str

kaapstorm committed:
Catch CaseNotFound

kaapstorm committed:
Merge pull request #29299 from dimagi/nh/fhir/rep-3

FHIRRepeater part 3 of 4

czue committed:
use constants

czue committed:
use more succinct sort

czue committed:
Merge pull request #29323 from dimagi/cz/messaging-tests

PR feedback on messaging tests

orangejenny committed:
Merge pull request #29322 from dimagi/jls/fix-queryData-clear

Fixed queryData not getting cleared

orangejenny committed:
Update staging.yaml

mjriley committed:
Merge pull request #29317 from dimagi/mjr/xss_translations

Removed mark_safe usages in translations

kaapstorm committed:
'updating files: scripts/staging.yaml'

'[ci skip]'

kaapstorm committed:
Merge branch 'master' into mk/fhir/api
